### PR TITLE
Rework owned sample

### PR DESCRIPTION
**Pull request by sashacmc**
Rework owned sample processing for pull and sub_channel:
 - Add multithreaded collections: `fifo_mt` and `ring_mt`
 - Add channel definition macro: `_Z_CHANNEL_DEFINE`
 - Add closure for `z_owned_sample_t`: `z_owned_closure_owned_sample_t` 
 - Define `z_owned_sample_ring_channel_t` and `z_owned_sample_fifo_channel_t` channels

_(NB: channels for queries and replies will be provided in the separate PR)_

### CODE REVIEW

File: examples/unix/c11/z_sub_channel.c
Comment (Mallets) [inline examples/unix/c11/z_sub_channel.c:67]: In zenoh-c we have `z_*_new()`: https://github.com/eclipse-zenoh/zenoh-c/blob/c1498fcaf08d3dad15e49b1ac15cd380c0dc175c/examples/z_queryable_with_channels.c#L57

We can decide to keep the new or to remove it. In any case, zenoh-pico and zenoh-c should be aligned.

File: include/zenoh-pico/api/handlers.h
Comment (Mallets) [inline include/zenoh-pico/api/handlers.h:68]: Should we have as well a `try_recv` method? It could be quite handy similar to what is being done in https://github.com/eclipse-zenoh/zenoh/pull/903/files.

It could come in a separate PR.

File: src/collections/fifo_mt.c
Comment (Mallets) [inline src/collections/fifo_mt.c:85]: Mainly a naming thing. `res` is actually `err` here, isn't it?

File: src/collections/fifo_mt.c
Comment (Mallets) [inline src/collections/fifo_mt.c:119]: Mainly a naming thing. `res` is actually `err` here, isn't it?

File: src/collections/fifo_mt.c
Comment (Mallets) [reply src/collections/fifo_mt.c:85]: The return type is `int8_t` but `res` has `int` type.

File: src/collections/ring_mt.c
Comment (Mallets) [inline src/collections/ring_mt.c:95]: The return type is `int8_t` but `res` has `int` type.

File: tests/z_channels_test.c
Comment (Mallets) [inline tests/z_channels_test.c:31]: Just a curiosity. Are you using here ` do { ... } while (0)` as a way to implement scoping `{ ... }`?

File: src/api/handlers.c
Comment (Mallets) [inline src/api/handlers.c:27]: In zenoh-pico we prefer explicit pointer checks, i.e. `if (dst != NULL && src != NULL)`

File: src/collections/fifo_mt.c
Comment (jean-roland) [inline src/collections/fifo_mt.c:28]: all the zp_* system calls are deprecated, you should use z_mutex & z_condvar instead

File: src/api/handlers.c
Comment (Mallets) [inline src/api/handlers.c:28]: If think we should check if `dst->_value != NULL` before doing the copy.

File: examples/unix/c11/z_sub_channel.c
Comment (sashacmc) [reply examples/unix/c11/z_sub_channel.c:67]: It seemed to me that I saw a creation without `_new`, but I now double-checked, this is not so, indeed in most cases we have a creation with `_new`, thanks!

File: src/collections/fifo_mt.c
Comment (sashacmc) [reply src/collections/fifo_mt.c:85]: Yes, err will be better, thanks!

File: tests/z_channels_test.c
Comment (jean-roland) [inline tests/z_channels_test.c:49]: With a closing `;` on the macro and another one on the call, `-Wpedantic` won't be happy

File: tests/z_channels_test.c
Comment (jean-roland) [reply tests/z_channels_test.c:31]: I explained the rationale: https://github.com/eclipse-zenoh/zenoh-pico/pull/307#pullrequestreview-1797921122

File: src/collections/fifo_mt.c
Comment (sashacmc) [reply src/collections/fifo_mt.c:28]: I don't see them in this branch :(

File: src/collections/fifo_mt.c
Comment (jean-roland) [inline src/collections/fifo_mt.c:26]: Still needs to replace the deprecrated systems calls: https://github.com/eclipse-zenoh/zenoh-pico/blob/main/src/deprecated/platform.c

Comment (jean-roland) [reply src/collections/fifo_mt.c:26]: Just realized this changes won't be in this branch until #394 is merged. So we either accept as is and modify later  or we wait till #394 is merged and rebase on that. Opinions @Mallets?

Comment (Mallets) [reply src/collections/fifo_mt.c:26]: I suggest to ago ahed with this PR and we then update #394. @jean-roland what do you think?

**Review by jean-roland — APPROVED** [review]

File: src/net/memory.c
Comment (Mallets) [inline src/net/memory.c:56]: So `dst->encoding` is left unitialized?

Comment (sashacmc) [reply src/net/memory.c]: Yes, this is not very good, I agree, I added a copy of the current encoding implementation

**Review by Mallets — APPROVED** [review]

**Pull request merged**
